KATANA_LOG_DEBUG_ASSERT -> KATANA_LOG_DEBUG_VASSERT

scober opened the pull request:
A few ASSERTS in the random access arrow builder would be more
informative for debugging if they were VASSERTS.

I made this change for my own debugging and thought I would share the love.

roshandathathri commented on libquery/include/querying/OpGraph.h:
```diff
@@ -8,31 +8,35 @@
   using OpPtr = std::shared_ptr<Operator>;
 
-  OpGraph() = default;
+  void SetOwner(std::shared_ptr<ExecutionPlan> plan) { owner_ = plan; }
```
What's the purpose for this?

bowu replied:
The OpGraph can access the execution context through its owner, an execution plan.

roshandathathri replied:
Why does OpGraph need to access the plan? It is sufficient for the plan to access the OpGraph right?

bowu replied:
This is just in case we need more than the data graph, but I can change the code to pass a pointer to OpGraph to access the execution context.

roshandathathri commented on libquery/include/querying/ops/MatchOp.h:
```diff
@@ -28,6 +28,9 @@ class MatchOp : public Operator {
 
 private:
+  //! Data graph representing the input data graph to the query
+  AttributedGraph data_graph_;
```
Why is this needed?

bowu replied:
Good catch. This is not needed, as MatchOp inherits a data graph from Operator. I'll remove it.

roshandathathri commented on libquery/include/querying/ops/Operator.h:
```diff
@@ -94,6 +101,8 @@ class Operator : public std::enable_shared_from_this<Operator> {
     children_.push_back(child);
   }
 
+  std::shared_ptr<AttributedGraph> data_graph_;
```
Why is this needed?

bowu replied:
If we don't have this, every time an op would like to access the data graph it has to obtain it from the execution context.

roshandathathri replied:
Data graph will be accessed during execution right?

bowu replied:
That's correct.

bowu commented:
Does anyone know why the Shared check fails but the Release check succeeds?

ddn0 commented on libquery/include/querying/OpGraph.h:
```diff
@@ -8,31 +8,35 @@
 
 private:
   std::vector<OpPtr> ops_;
+  std::shared_ptr<AttributedGraph> data_graph_{nullptr};
```
FWIW, I think we should be wary of `shared_ptr` everywhere. It obscures when objects actually stopped being referenced.

In general it would be better to just have a clear container object like the plan, and then objects can be tied to the lifetime of the container object and they can simply borrow references (T*) to parts of the plan.

l-hoang replied:
In the case of the data_graph_, having a unique pointer for the Context (the owner) and having the children of Context keep a pointer to it is what I would suggest.

bowu replied:
Make sense.

bowu replied:
The reason that I used shared_ptr is because I wanted to avoid two indirections to get it from an ExecutionContext. We can revisit this later.

l-hoang replied:
I think what we're saying is that you don't need to keep it a shared_ptr but rather a regular pointer AttributedGraph; it  avoids the additional reference increment to the shared pointer.

bowu replied:
Got it.

ddn0 commented on libquery/include/querying/QueryServer.h:
```diff
@@ -18,17 +18,18 @@ class QueryServer {
-    // TODO(bowu) plan/query graph is tied to some schema; query_str is not
-    // enough, need graph schema to create it; tie graph schema to cache as well
-    std::shared_ptr<ExecutionPlan> plan = cache_.GetPlan(query_str);
+    // TODO: this is a temporary fix; graph may change to make the plan invalid
```
probably keep your name associated with TODOs.

ddn0 commented on libquery/include/querying/ops/Operator.h:
```diff
@@ -7,9 +7,12 @@
 
 namespace galois {
 
+using AttributedGraph = graphs::AttributedGraph;
```
ditto

ddn0 commented on libquery/include/querying/ops/Operator.h:
```diff
@@ -7,9 +7,12 @@
 
+using AttributedGraph = graphs::AttributedGraph;
+
 class Operator : public std::enable_shared_from_this<Operator> {
```
For graph-like data structures, I wouldn't expect reference counting to be a great fit (even if you can assert that it will be a DAG). I'd opt for a container that owns all the Ops in a graph.

ddn0 commented on libquery/CMakeLists.txt:
```diff
@@ -16,6 +16,7 @@ target_sources(galois_query PRIVATE
   src/Matching.cpp
   src/RegularPathQuery.cpp
   src/SubgraphQuery.cpp
+  src/OpGraph.cpp
```
small nit: just sort this list and keep insertions sorted.

l-hoang replied:
Is there an automated way to do this sorting or will it have to be good ol' eyeballing and line moving?

ddn0 replied:
in vim, I do a visual select followed by `:!sort`

ddn0 commented on libquery/src/OpGraph.cpp:
```diff
@@ -0,0 +1,18 @@
+  if (!plan) {
+    GALOIS_LOG_FATAL("OpGraph does not have an owner");
+  }
+  auto dataGraph = plan->GetExecutionContext().GetDataGraph();
```
`data_graph`

l-hoang commented on libquery/include/querying/ExecutionPlan.h:
```diff
@@ -1,37 +1,36 @@
-  ExecutionPlan(std::string query_str,
-                const galois::BitLabelSchema& bit_label_schema)
-      : query_str_(query_str) {
+  ExecutionPlan(ExecutionContext ctx) : ctx_(ctx) {
```
Pass this in via reference to avoid a copy. `ExecutionContext&`

l-hoang commented on libquery/include/querying/ExecutionPlan.h:
```diff
@@ -59,11 +58,9 @@ class ExecutionPlan {
   }
 
-  std::string query_str_;
+  ExecutionContext ctx_;
```
Make this a reference, though my assumption here is that the context object is being kept live somewhere.

My guess is that you will reuse the context in many plans though.

l-hoang commented on libquery/include/querying/ExecutionContext.h:
```diff
@@ -0,0 +1,33 @@
+  unsigned int GetOptLevel() { return opt_level_; }
+
+private:
+  std::string query_str_;
```
Each context has its own query?

bowu commented on libquery/include/querying/ExecutionContext.h:
```diff
@@ -0,0 +1,33 @@
+
+namespace galois {
+
+using AttributedGraph = graphs::AttributedGraph;
```
Will do.

bowu commented:
@ddn0 Can you take a look at the status checks? It seems a mac os check failure stopped the world.

l-hoang commented on libquery/include/querying/OpGraph.h:
```diff
@@ -4,35 +4,37 @@
 namespace galois {
 
-using AttributedGraph = graphs::AttributedGraph;
+class ExecutionContext;
```
Prefer to include the header to the Context class rather than declare a forward reference.

l-hoang commented on libquery/include/querying/ops/FilterOp.h:
```diff
@@ -6,7 +6,7 @@ namespace galois {
 class FilterOp : public Operator {
 public:
-  virtual void Compute() final {}
+  std::vector<Operator*> Compute() final { return PropagateOutput(); }
```
I'm assuming this is dummy work for now (as in this isn't the final Compute function).

l-hoang commented on libquery/include/querying/ops/Operator.h:
```diff
@@ -73,39 +75,56 @@ class Operator : public std::enable_shared_from_this<Operator> {
     }
+    // clear the output if it's not a terminal op
+    if (NumChildren() != 0) {
+      output_ = nullptr;
```
output_.reset() is an explicit release of the pointer and makes it clear that this is a shared pointer.

Commit:
KATANA_LOG_DEBUG_ASSERT -> KATANA_LOG_DEBUG_VASSERT

A few ASSERTS in the random access arrow builder would be more
informative for debugging if they were VASSERTS.

arthurp reviewed (approved):
Always good to document things.
